thali/thalireplicationmanager.js

```diff
@@ -0,0 +1,301 @@
+'use strict';
+
+var ThaliEmitter = require('./thaliemitter');
+var EventEmitter = require('events').EventEmitter;
+var inherits = require('util').inherits;
+var net = require('net');
+var multiplex = require('multiplex');
+var validations = require('./validations');
+
+var e = new EventEmitter();
+
+var PEER_AVAILABILITY_CHANGED = ThaliEmitter.events.PEER_AVAILABILITY_CHANGED;
+var NETWORK_CHANGED = ThaliEmitter.events.NETWORK_CHANGED;
+
+inherits(ThaliReplicationManager, EventEmitter);
+
+function ThaliReplicationManager(db, emitter) {
+  this._db = db;
+  this._emitter = (emitter || new ThaliEmitter());
+  this._replications = {};
+  this._clients = {};
+  this._isStarted = false;
+  this._serverBridge = null;
+  this._serverBridgePort = 0;
+  EventEmitter.call(this);
+}
+
+ThaliReplicationManager.events = {
+  STARTING: 'starting',
+  STARTED: 'started',
+  STOPPING: 'stopping',
+  STOPPED: 'stopped',
+  START_ERROR: 'startError',
+  STOP_ERROR: 'stopError',
+  CONNECT_ERROR: 'connectError',
+  DISCONNECT_ERROR: 'disconnectError',
+  SYNC_ERROR: 'syncRetry'
+};
+
+/**
+* Starts the Thali replication manager with the given device name and port number
+* @param {String} deviceName the device name to advertise.
+* @param {Number} port the port number used for synchronization.
+*/
+ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
+  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+  validations.ensureValidPort(port);
+  validations.ensureNonNullOrEmptyString(dbName, 'dbName');
+
+  this.emit(ThaliReplicationManager.events.STARTING);
+
+  this._port = port;
+  this._deviceName = deviceName;
+  this._dbName = dbName;
+  this._serverBridge = muxServerBridge.call(this, port);
```

Comment:
If someone calls start twice won't we end up leaking a serverBridge? Shouldn't we check to see if start has already been called without calling stop and return an error? That would prevent the leak.